src/common/snippets/src/lowered/pass/extract_loop_invariants.cpp

```diff
@@ -0,0 +1,204 @@
+        const auto& inner_loop_input_ports = inner_loop_info->get_input_ports();
+        const auto& potential_extractable_exprs = get_potential_extractable_exprs(inner_loop_input_ports);
+        bool has_been_extraction = false;
+        for (const auto& port_expr : potential_extractable_exprs) {
+            if (is_extraction_applicable(port_expr, inner_loop_info)) {
+                status = true;
+                LinearIR::constExprIt inner_loop_begin_pos, inner_loop_end_pos;
+                std::tie(inner_loop_begin_pos, inner_loop_end_pos) =
+                    loop_manager->get_loop_bounds(linear_ir, inner_loop_id);
+                // extract scalar on inputs if there are
+                for (size_t i = 0; i < port_expr->get_input_count(); ++i) {
+                    auto parent = port_expr->get_input_port_connector(i)->get_source().get_expr();
+                    if (ov::is_type<snippets::op::Scalar>(parent->get_node())) {
+                        extract_expr(parent, linear_ir, inner_loop_begin_pos, inner_loop_end_pos);
+                    }
+                }
+                extract_expr(port_expr, linear_ir, inner_loop_begin_pos, inner_loop_end_pos);
+                update_loop_ports(port_expr, loop_manager, inner_loop_id, inner_loop_begin_pos, inner_loop_end_pos);
+                has_been_extraction = true;
+                break;  // extracted and refreshed loop_input_ports. break potential_extractable_exprs loop, and go while() to start again.
+            }
+        }
+        if (inner_loop_input_ports.size() == 0 && inner_loop_info->get_output_ports().size() == 0) {
+            // become a empty(inner_loop_input_ports is ref) loop after extraction, let remove it from loop_manager
+            loop_manager->remove_loop_info(inner_loop_id);
+            extraction_completed = true;
+            break;
+        }
+        // no more extractable expr in this loop after for() of iter loop ports, done for this loop.
+        if (!has_been_extraction)
+            extraction_completed = true;
+    }
+    return status;
+}
+
+bool ExtractLoopInvariants::run(LinearIR& linear_ir, lowered::LinearIR::constExprIt begin, lowered::LinearIR::constExprIt end) {
+    OV_ITT_SCOPED_TASK(ov::pass::itt::domains::SnippetsTransform, "Snippets::ExtractLoopInvariants")
+    bool modified = false;
+
+    // move invariant expr to top(outside) of current loop
+    for (auto expr_it = begin; expr_it != end; expr_it++) {
+        const auto& next_expr_it = std::next(expr_it);
+        if (next_expr_it == linear_ir.end())
+            break;
+        const auto& expr = *expr_it;
+        const auto& next_expr = *next_expr_it;
+        const auto& current_loop_ids = expr->get_loop_ids();
+        const auto& next_loop_ids = next_expr->get_loop_ids();
+        const auto& current_loop_ids_size = current_loop_ids.size();
+        const auto& next_loop_ids_size = next_loop_ids.size();
+
+        if (next_loop_ids_size < current_loop_ids_size) {
```

Comment:
What if `next_loop_ids_size == current_loop_ids_size`? The expr can still be the last in the inner loop:
expr1 loop_ids : [1,]
expr2 loop_ids : [1,2]
expr3 loop_ids : [1,3]
expr4 loop_ids : [1]
how will we handle such cases here? Can we extract from loop_id = 2?

Reply:
I used to think about it, and did not cover this because in my opinion there should be an `intermediate_buffer `between loop 2 and loop 3. In real case it should be:
expr2 loop_ids :           [1,2]
expr(buffer) loop_ids : [1]
expr3 loop_ids :           [1,3]
correct me if I missed something please. Thanks Ivan!

Reply:
You are right that in present state of the pipeline the Buffer is likely to be there, but necessarily. For example, if expressions in loops 2 & 3 work directly with inputs&output memory then the Buffers will not be needed.
Anyways, we really should try to write the transformations as generic as possible and to keep the number of constraints and assumptions minimal. 

From the semantics of this transformation, I don't really see why shouldn't we support the example I provided. So let's try to cover such cases as well. If you think that it's too complicated (or we don't have enough time) to do it in this PR, let's create a follow-up ticket and finish it in the next release cycle.

In general, we are trying to build modular and scalable infrastructure. So we should do our best to develop robust and generic transformations. Even in this PR, you encountered a problem with the `NormalizeLoopIDs` pass because the pass assumed that the LoopEnd expressions are present in the gap (which is not really needed to normalize loop ids). That's exactly the kind of problems we are trying to avoid.
And as I said, feel free to create a ticket for the `NormalizeLoopIDs` generalization

Reply:
I have implemented  a generic `lood_id identification` method, witch cover parallel loop on the same loop depth. Could you please take a look :-) [2840f72](https://github.com/openvinotoolkit/openvino/pull/24044/commits/2840f72b43c4adc600711daa407aa919159b9cf8)

Reply:
@IvanNovoselov , This is the more generic method to generate loops in order(extraction order is from inner to outer) to extract: [31e5563](https://github.com/openvinotoolkit/openvino/pull/24044/commits/31e5563217b38f19ef6e5437cb7e6aa68d03ffb3)